From be37850200fe86284c98881d9e30b34ccbb35855 Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Wed, 18 Jan 2023 10:57:37 +0800
Subject: [PATCH] mm: oom_kill: fix KABI broken by "oom_kill.c: futex: delay
 the OOM reaper to allow time for proper futex cleanup"
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: be37850200fe86284c98881d9e30b34ccbb35855
Modified-by-SEL: No


hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I61FDP
CVE: NA

-------------------------------

Move oom_reaper_timer from task_struct to task_struct_resvd to fix KABI
broken.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
Reviewed-by: Nanyong Sun <sunnanyong@huawei.com>
Reviewed-by: chenhui <judy.chenhui@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 include/linux/sched.h |  5 ++++-
 mm/oom_kill.c         | 11 ++++++-----
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index afc7651a2659..d748c6f16174 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -675,6 +675,10 @@ struct wake_q_node {
 struct task_struct_resvd {
 	/* pointer back to the main task_struct */
 	struct task_struct	*task;
+
+#ifdef CONFIG_MMU
+	struct timer_list	oom_reaper_timer;
+#endif
 };
 
 struct task_struct {
@@ -1353,7 +1357,6 @@ struct task_struct {
 	int				pagefault_disabled;
 #ifdef CONFIG_MMU
 	struct task_struct		*oom_reaper_list;
-	struct timer_list		oom_reaper_timer;
 #endif
 #ifdef CONFIG_VMAP_STACK
 	struct vm_struct		*stack_vm_area;
diff --git a/mm/oom_kill.c b/mm/oom_kill.c
index 2933e2beba6f..7eb4fda1ce87 100644
--- a/mm/oom_kill.c
+++ b/mm/oom_kill.c
@@ -706,8 +706,9 @@ static int oom_reaper(void *unused)
 
 static void wake_oom_reaper(struct timer_list *timer)
 {
-	struct task_struct *tsk = container_of(timer, struct task_struct,
-			oom_reaper_timer);
+	struct task_struct_resvd *tsk_resvd = container_of(timer,
+			struct task_struct_resvd, oom_reaper_timer);
+	struct task_struct *tsk = tsk_resvd->task;
 	struct mm_struct *mm = tsk->signal->oom_mm;
 	unsigned long flags;
 
@@ -741,9 +742,9 @@ static void queue_oom_reaper(struct task_struct *tsk)
 		return;
 
 	get_task_struct(tsk);
-	timer_setup(&tsk->oom_reaper_timer, wake_oom_reaper, 0);
-	tsk->oom_reaper_timer.expires = jiffies + OOM_REAPER_DELAY;
-	add_timer(&tsk->oom_reaper_timer);
+	timer_setup(&tsk->_resvd->oom_reaper_timer, wake_oom_reaper, 0);
+	tsk->_resvd->oom_reaper_timer.expires = jiffies + OOM_REAPER_DELAY;
+	add_timer(&tsk->_resvd->oom_reaper_timer);
 }
 
 static int __init oom_init(void)
-- 
2.33.0

